nuxeo-mongodb-directory/src/main/java/org/nuxeo/directory/MongoDBSession.java

```diff
@@ -0,0 +1,349 @@
+        checkPermission(SecurityConstants.WRITE);
+        checkDeleteConstraints(id);
+        try {
+            DeleteResult result = getCollection().deleteOne(
+                    MongoDBSerializationHelper.fieldMapToBson(Collections.singletonMap(getIdField(), id)));
+            if (!result.wasAcknowledged()) {
+                throw new DirectoryException("Error while deleting the entry");
+            }
+        } catch (MongoWriteException e) {
+            throw new DirectoryException(e);
+        }
+
+    }
+
+    @Override
+    public void deleteEntry(String id, Map<String, String> map) throws DirectoryException {
+        deleteEntry(id);
+    }
+
+    @Override
+    public DocumentModelList query(Map<String, Serializable> filter) throws DirectoryException {
+        return query(filter, Collections.emptySet());
+    }
+
+    @Override
+    public DocumentModelList query(Map<String, Serializable> filter, Set<String> fulltext) throws DirectoryException {
+        return query(filter, fulltext, new HashMap<>());
+    }
+
+    @Override
+    public DocumentModelList query(Map<String, Serializable> filter, Set<String> fulltext, Map<String, String> orderBy)
+            throws DirectoryException {
+        return query(filter, fulltext, orderBy, false);
+    }
+
+    @Override
+    public DocumentModelList query(Map<String, Serializable> filter, Set<String> fulltext, Map<String, String> orderBy,
+            boolean fetchReferences) throws DirectoryException {
+        return query(filter, fulltext, orderBy, fetchReferences, -1, -1);
+    }
+
+    @Override
+    public DocumentModelList query(Map<String, Serializable> filter, Set<String> fulltext, Map<String, String> orderBy,
+            boolean fetchReferences, int limit, int offset) throws DirectoryException {
+
+        Document bson = buildQuery(filter, fulltext);
+
+        DocumentModelList entries = new DocumentModelListImpl();
+
+        FindIterable<Document> results = getCollection().find(bson).skip(offset);
+        if (limit > 0) {
+            results.limit(limit);
+        }
+        for (Document resultDoc : results) {
+            // Cast object to document model
+            Map<String, Object> fieldMap = MongoDBSerializationHelper.bsonToFieldMap(resultDoc);
+            entries.add(fieldMapToDocumentModel(fieldMap));
+        }
+
+        if (orderBy != null && !orderBy.isEmpty()) {
+            getDirectory().orderEntries(entries, orderBy);
+        }
+
+        return entries;
+    }
+
+    protected Document buildQuery(Map<String, Serializable> fieldMap, Set<String> fulltext) {
+
+        Document bson = new Document();
+        for (Map.Entry<String, Serializable> entry : fieldMap.entrySet()) {
+            Object value = MongoDBSerializationHelper.valueToBson(entry.getValue());
+            if (value != null) {
+                if (fulltext.contains(entry.getKey())) {
+                    switch (substringMatchType) {
```

Comment:
Note that `// TODO this should be case-insensitive` and add an `@Ignore` unit test that fails to show that we need to implement it.